cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/BodyConsumerAdapter.java

```diff
@@ -0,0 +1,195 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.service.http;
+
+import co.cask.cdap.api.Transactional;
+import co.cask.cdap.api.TxRunnable;
+import co.cask.cdap.api.data.DatasetContext;
+import co.cask.cdap.api.service.http.HttpContentConsumer;
+import co.cask.cdap.api.service.http.HttpServiceResponder;
+import co.cask.cdap.api.service.http.NonTransactionalHttpContentConsumer;
+import co.cask.cdap.common.lang.ClassLoaders;
+import co.cask.cdap.common.lang.CombineClassLoader;
+import co.cask.http.BodyConsumer;
+import co.cask.http.HttpResponder;
+import co.cask.tephra.TransactionContext;
+import co.cask.tephra.TransactionFailureException;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Multimap;
+import com.ning.http.client.HttpContent;
+import org.jboss.netty.buffer.ChannelBuffer;
+
+import java.nio.ByteBuffer;
+import javax.annotation.Nullable;
+
+/**
+ * An adapter class to delegate calls from {@link BodyConsumer} to {@link HttpContentConsumer}.
+ */
+final class BodyConsumerAdapter extends BodyConsumer {
+
+  private final DelayedHttpServiceResponder responder;
+  private final HttpContentConsumer delegate;
+  private final Transactional transactional;
+  private final ClassLoader contextClassLoader;
+
+  /**
+   * Constructs a new instance.
+   *
+   * @param responder the responder used for sending response back to client
+   * @param delegate the {@link HttpContentConsumer} to delegate calls to
+   * @param datasetContext context for accessing dataset inside transaction
+   * @param txContext the {@link TransactionContext} for creating transaction
+   */
+  BodyConsumerAdapter(DelayedHttpServiceResponder responder, HttpContentConsumer delegate,
+                      DatasetContext datasetContext, TransactionContext txContext) {
+    this.responder = responder;
+    this.delegate = delegate;
+    this.transactional = createTransactional(txContext, datasetContext);
+    this.contextClassLoader = new CombineClassLoader(null, ImmutableList.of(delegate.getClass().getClassLoader(),
+                                                                            getClass().getClassLoader()));
+  }
+
+  @Override
+  public void chunk(final ChannelBuffer request, HttpResponder responder) {
+    try {
+      // Based on the type of the content consumer, call onReceive with transaction or not.
+      if (delegate instanceof NonTransactionalHttpContentConsumer) {
+        final ClassLoader oldClassLoader = ClassLoaders.setContextClassLoader(contextClassLoader);
```

Comment:
First, `contextClassLoader` is not the program class loader, but rather a combine class loader that prefer using program class loader to load class first (we do this for all program execution context, otherwise for program that uses HBase directly, it cannot find the right resource for the HBaseConfiguration). Anyway, the `contextClassLoader` is the class loader needs to be set as the context class loader whenever we are going to invoke user method.

The `oldClassLoader` is the existing context class loader.

We need the switching because we are going to call the `NonTransactionalHttpContentConsumer.onReceive` method, which is a user method.

The `Transactional